ntlm: use strict order for SSL backend #if branches

vszakats opened the pull request:
With the recently introduced MultiSSL support multiple SSL backends
can be compiled into cURL That means that now the order of the SSL

One option would be to use the same SSL backend as was configured
via `curl_global_sslset()`, however, NTLMv2 support would appear
to be available only with some SSL backends. For example, when
eb88d778e (ntlm: Use Windows Crypt API, 2014-12-02) introduced
support for NTLMv1 using Windows' Crypt API, it specifically did
*not* introduce NTLMv2 support using Crypt API at the same time.

So let's select one specific SSL backend for NTLM support when
compiled with multiple SSL backends, using a priority order such
that we support NTLMv2 even if only one compiled-in SSL backend can
be used for that.

---

metalink: fix NSS issue in MultiSSL builds

In MultiSSL mode (i.e. when more than one SSL backend is compiled
in), we cannot use the compile time flag `USE_NSS` as indicator that
the NSS backend is in use. As far as Metalink is concerned, the SSL
backend is only used for MD5, SHA-1 and SHA-256 calculations,
therefore one of the available SSL backends is selected at compile
time, in a strict order of preference.

Let's introduce a new `HAVE_NSS_CONTEXT` constant that can be used
to determine whether the SSL backend used for Metalink is the NSS
backend, and use that to guard the code that wants to de-initialize
the NSS-specific data structure.

dscho commented:
The rationale appears sound to me, and as far as I can see the patch really only reorders the clauses into a more sensible order.

I have not been able to find time to wrap my head around the ramifications of using only one SSL backend's capabilities even when multiple SSL backends are compiled in, though.

I would have expected more of a patch in line with 69039fd1fa3e65e2f5737b2a4044dcb8fbabb76d, to be honest, i.e. moving those SSL backend-specific functions into the respective backends' source files and accessing them via `Curl_ssl->[function-name]`. In fact, I would much prefer that, as it would ensure that those functions really mesh well with the currently-active SSL backend.

@vszakats what do you think?

vszakats commented:
@dscho I was looking into the NTLM source and it seems that what it uses is _crypto primitives_ (MD4, HMAC-MD5, DES-ECB), and not much (if anything) TLS/SSL-related. Such crypto primitives should be interchangeable from a user's POV (except possible security bugs and speed, which is easy to tweak by prioritizing them), so unless I missed something else, tapping on the dynamic backend function table wouldn't give a big benefit here. One more question remains: If this observation is true, how is it possible that OpenSSL/NSS cannot support _some_ NTLM features?

dscho commented:
Hmm. I am *far* from an expert in things NTLM, let alone cURL's NTLM code. But I think you have a point there: if OpenSSL/NSS have limitations re: NTLM, then it probably *does* make a difference which crypto primitives you use...

vszakats commented:
Re-reading the original commit message [1] and your comment, it seems I got the message wrong. It says that OpenSSL and NSS are the _superior_ choice, and built-in Windows Crypt comes short. This is no less interesting and doesn't answer the "why". Also, Windows Crypt was the last backend checked, so now I'm lost what bd2231104be563684aa48cd43ab2ed2d1ee4c5b4 was attempting to fix.

I'm going to reorder the backends, moving OpenSSL and NSS back to the top and WinSSL at the end, largely reverting to the original order.

vszakats commented:
Reworked accordingly. This sorts out the OpenSSL vs NSS vs Windows Crypto issue from this PR.

@dscho: This [patch](https://github.com/curl/curl/pull/1840#issuecomment-326087305) of yours seems safe to apply regardless of this PR.

jay commented:
You know I would think if anything can handle NTLM properly it would be Windows crypto. It would be nice to know what particular bug caused that change.

/cc @captain-caveman2k

dscho commented:
> @dscho: This patch of yours seems safe to apply regardless of this PR.

I pushed it to my fork as [`ntlm-fixes`](https://github.com/curl/curl/compare/master...dscho:ntlm-fixes), but I want to make sure that only the initialized SSL backend's routines are used. That really is required, [as reported by Kamil Dudka on the mailing list](https://curl.haxx.se/mail/lib-2017-08/0124.html).

vszakats commented:
@dscho The issue reported by @kdudka seems to be caused not by this particular source file (`curl_ntlm_core.c`), but by `vauth/ntlm.c`, which has a single `USE_NSS` branch that doesn't respect the strict order of the SSL backends as document/fixed in this patch. IOW the NSS-specific code gets activated whenever NSS is enabled, while it should only be active when OpenSSL and GnuTLS are not available.

This patch should fix it (I haven't tested it though):
```patch
diff --git a/lib/vauth/ntlm.c b/lib/vauth/ntlm.c
index 42196455f..e9db156f4 100644
--- a/lib/vauth/ntlm.c
+++ b/lib/vauth/ntlm.c
@@ -44,7 +44,10 @@
 #include "rand.h"
 #include "vtls/vtls.h"
 
-#ifdef USE_NSS
+#if !defined(USE_OPENSSL) && \
+    !defined(USE_GNUTLS_NETTLE) && \
+    !defined(USE_GNUTLS) && \
+    defined(USE_NSS)
 #include "vtls/nssg.h" /* for Curl_nss_force_init() */
 #endif
 
@@ -272,7 +275,10 @@ CURLcode Curl_auth_decode_ntlm_type2_message(struct Curl_easy *data,
   unsigned char *type2 = NULL;
   size_t type2_len = 0;
 
-#if defined(USE_NSS)
+#if !defined(USE_OPENSSL) && \
+    !defined(USE_GNUTLS_NETTLE) && \
+    !defined(USE_GNUTLS) && \
+    defined(USE_NSS)
   /* Make sure the crypto backend is initialized */
   result = Curl_nss_force_init(data);
   if(result)
```

If it does resolve the test failure, it could be added to this PR. This would make dynamically switching crypto-primitives (and the complications that come with it) unnecessary.

dscho commented:
> ```diff
> -#if defined(USE_NSS)
> +#if !defined(USE_OPENSSL) && \
> +    !defined(USE_GNUTLS_NETTLE) && \
> +    !defined(USE_GNUTLS) && \
> +    defined(USE_NSS)
> /* Make sure the crypto backend is initialized */
> result = Curl_nss_force_init(data);
> ```

Actually, think you want to keep the original guard for the `#include` and add a *runtime guard* here, i.e.

```diff
 #ifdef USE_NSS
-/* Make sure the crypto backend is initialized */
-result = Curl_nss_force_init(data);
+if (Curl_ssl_backend() == CURLSSLBACKEND_NSS) {
+  /* Make sure the crypto backend is initialized */
+  result = Curl_nss_force_init(data);
...
```

Rationale: you want to use the NSS routines only when NSS was selected as SSL backend. This is now a runtime switch, when it had been a compile time switch before.

vszakats commented:
Doing it dynamically adds runtime/size overhead [*], so until we know for certain there is a benefit in runtime switching between crypto primitives offered by different SSL backends in context of NTLM, I'd rather stay on the course to select the most ideal available backend at compile-time.

[*] Adding a runtime guard here alone would only be useful when OpenSSL/GnuTLS are not available and `curl_ntlm_core.c` would actually call into NSS. To make the call universally useful, the whole `curl_ntlm_core.c` would also need to be reworked to do similar runtime branching.

dscho commented:
> I'd rather stay on the course to select the most ideal available backend at compile-time.

Makes sense.

Are you taking it from here? Or should I?

vszakats commented:
I can add the `ntlm.c` patch to this PR and unless there is new information on backend priority, this PR seems to cover the issue (though it'd be nice if @kdudka could retest - I'm lacking any sort of facility for actual NTLM tests). If it doesn't, or this isn't enough, I don't mind at all if you take it over :)

dscho commented:
I was about to ask @kdudka to retest, but while composing my reply to his mail, I realized that he provided us with a nice command-line to test the thing ourselves:

```sh
/buildconf &&
./configure --with-ssl --with-nss \
        --with-default-ssl-backend=openssl --disable-tls-srp &&
make -j16 &&
cd tests &&
./runtests.pl -a -p -v "HTTP NTLM auth"
```

And it does indeed fail here with

```
...
TESTFAIL: These test cases failed: 176 2025 2028 2029 2030 2031 2032 2033
...
```

If I add [that runtime check I proposed earlier](https://github.com/curl/curl/pull/1848#issuecomment-327766942), the test 176 is fixed (but not the others).

I pushed the proposed fix to my fork, please feel free to integrate the commit via `git fetch https://github.com/dscho/curl ntlmmulti` into your branch.

Will try to figure out what goes wrong with the remaining seven test failures...

dscho commented:
> Will try to figure out what goes wrong with the remaining seven test failures...

A ha! I forgot to re-`make` in `tests/`. After I remembered, all of the test cases that were previously failing now pass!

So yes, we really need that runtime check, apparently.

vszakats commented:
It'd interesting the know _why_ NSS needs to be initialized when OpenSSL is supposed to be used. Maybe some NSS-specific code is still lurking somewhere in the NTLM codepath.

Once I successfully convince the build to correctly recognize the NSS package on darwin/Homebrew, I hope to run the test myself - thanks for the formula!

dscho commented:
> So yes, we really need that runtime check, apparently.

Just to make sure that I get my point across: if we *always* force-init NSS, even when the current backend is OpenSSL, the NTLM authentication *will fail*. If we force-init NSS only when the current backend is NSS, it will succeed.

dscho commented:
> It'd interesting the know why NSS needs to be initialized when OpenSSL is supposed to be used.

Actually, the opposite is true: when OpenSSL is supposed to be used, NSS *must not be initialized*. That is what my commit ensures: https://github.com/dscho/curl/commit/53155a4782ea64c85a7e8fa1928155def32644bf

vszakats commented:
Hm, that's what is my `ntlm.c` patch was intended to do, but at compile-time: calling `Curl_nss_force_init()` only when OpenSSL is _not_ used.

From your patch it's clear that there is another location where the NSS initialization is done: `lib/http_ntlm.c`. Patch below does the same trick for this file:
```patch
diff --git a/lib/http_ntlm.c b/lib/http_ntlm.c
index 8a78bd293..bda209d48 100644
--- a/lib/http_ntlm.c
+++ b/lib/http_ntlm.c
@@ -41,7 +41,10 @@
 #include "vauth/vauth.h"
 #include "url.h"
 
-#if defined(USE_NSS)
+#if !defined(USE_OPENSSL) && \
+    !defined(USE_GNUTLS_NETTLE) && \
+    !defined(USE_GNUTLS) && \
+    defined(USE_NSS)
 #include "vtls/nssg.h"
 #elif defined(USE_WINDOWS_SSPI)
 #include "curl_sspi.h"
@@ -129,7 +132,10 @@ CURLcode Curl_output_ntlm(struct connectdata *conn, bool proxy)
   DEBUGASSERT(conn);
   DEBUGASSERT(conn->data);
 
-#ifdef USE_NSS
+#if !defined(USE_OPENSSL) && \
+    !defined(USE_GNUTLS_NETTLE) && \
+    !defined(USE_GNUTLS) && \
+    defined(USE_NSS)
   if(CURLE_OK != Curl_nss_force_init(conn->data))
     return CURLE_OUT_OF_MEMORY;
 #endif
```

dscho commented:
> Hm, that's what is my ntlm.c patch was intended to do, but at compile-time: calling Curl_nss_force_init() only when OpenSSL is not used.

But you cannot detect at compile-time what is used at runtime.

From my tests, it seemed that the `Curl_nss_force_init()` call interfered with an OpenSSL backend, but not with an NSS backend. That suggests to me that some NSS code is used in NTLM negotiations when NSS is configured as backend.

Therefore, I would prefer to go with the runtime check, also to avoid nasty surprises when NSS is compiled in with some other backend that is neither GNU TLS nor OpenSSL, hence unhandled by your compile-time switch.

dscho commented on lib/vauth/ntlm.c:
```diff
@@ -272,7 +275,10 @@ CURLcode Curl_auth_decode_ntlm_type2_message(struct Curl_easy *data,
+#if !defined(USE_OPENSSL) && \
+    !defined(USE_GNUTLS_NETTLE) && \
+    !defined(USE_GNUTLS) && \
+    defined(USE_NSS)
```
I am not sure that this is robust.

Granted, you switched the order of the conditionals above, and that happens to favor only OpenSSL and GNU TLS over NSS right now.

But you already have *two* other locations where you have to replicate that order partially: this place here and `lib/http_ntlm.c".

And I bet not even you would remember to update these in, say, 6 months from now, should another GNUTLS variant (or similar) be added.

So I still would prefer the runtime check, as it is a *lot* more robust.

Having said that, I am not a core cURL reviewer, therefore my opinion is a bit less important than e.g. @bagder's.

vszakats replied:
By 'used', I meant 'compiled-in'. Robustness can be argued of course, and most likely it can be made cleaner by moving these checks to a common header or other ways to ensure the correct order.

vszakats replied:
Moving this single check to runtime don't seem correct to me. Moving all checks to runtime seems overkill and bloaty. Though I won't have the resources for a long fight over this, so I'm closing this PR for now. If somebody will still be interested in this, please feel free to use it.

jay replied:
Let's revisit this, What about putting something like this in vtls.h
~~~c
/* If NTLM is the first available SSL backend (see order in curl_ntlm_core)
   then it must be initialized to be used by NTLM. */
#if !defined(USE_OPENSSL) && \
    !defined(USE_GNUTLS_NETTLE) && \
    !defined(USE_GNUTLS) && \
    defined(USE_NSS)
#define NTLM_NEEDS_NSS_INIT
#endif
~~~
then the other two files ntlm.c and http_ntlm.c use that define and change the comment block at the beginning to something like
~~~c
/* SSL backend-specific #if branches in this file must be kept in the order
   documented in curl_ntlm_core. */
~~~
this way we get rid of the duplication

vszakats replied:
@jay: It doesn't feel out-of-place this way, indeed, thanks for the suggestion. Added a commit that implements it.

dscho commented:
> also to avoid nasty surprises when NSS is compiled in with some other backend that is neither GNU TLS nor OpenSSL

I just realized how you avoided that for now: you changed the order of the `#if ... #elif ... #elif ...` chain in `curl_ntlm_core.c` so that it starts with `OPENSSL`, `GNUTLS_NETTLE`, `GNUTLS` and `NSS`. Meaning that NSS is favored for NTLM operations as long as neither OpenSSL nor GNU TLS support was compiled in.

As a consequence, you can get away here and in `vauth/ntlm.c` with checking only for `!OPENSSL && !GNUTLS_NETTLE && !GNUTLS && NSS`.

In fact, my runtime check (even if I would prefer it, still), would be *wrong* because NTLM *ignores* the current SSL backend and uses whatever comes first in the `if ... #elif ... #elif ...` chain and is compiled in.

But you would run into trouble if, say, NSS and SCHANNEL was compiled in and SCHANNEL was the current backend because then you still would want to use NSS for NTLM, but `Curl_nss_init()` would not have been called, hence `nss_initlock` would be uninitialized and the call to `Curl_nss_force_init()` would result in a nasty "unable to initialize NSS, curl_global_init() should have been called with CURL_GLOBAL_SSL or CURL_GLOBAL_ALL" error message.

dscho commented:
@vszakats oh, I just saw that you closed this... why?

vszakats commented:
I felt this goes way more complicated that I initially thought. But, not to worry, it can be reopened if a solution might still be in sight.

The problem with compile-time ordering is indeed that there may be other order-dependent parts in the source code and digging these out, ensuring them not to reappear and to keep them in the correct order might be an arduous task. Can't really judge the number of these hot-spots in the code, if low, this may work, if not, it's not practical.

vszakats commented:
I pushed the `ntlm.c` check to the branch (after closing, so it's missing from this page).
https://github.com/vszakats/curl/tree/ntlmmulti

vszakats commented:
A quick grep didn't reveal too many `USE_(NSS|OPENSSL|GNUTLS)` checks (besides the files touched in this PR and backend-specific VTLS sources):
```
./lib/curl_des.c:#if defined(USE_NTLM) && !defined(USE_OPENSSL)
./lib/curl_des.c:#endif /* USE_NTLM && !USE_OPENSSL */
./lib/hostcheck.c:#if defined(USE_OPENSSL)                                \
./lib/md4.c:#if defined(USE_NSS) || defined(USE_OS400CRYPTO) || \
./lib/md4.c:#endif /* defined(USE_NSS) || defined(USE_OS400CRYPTO) ||
./lib/md5.c:#if defined(USE_GNUTLS_NETTLE)
./lib/md5.c:#elif defined(USE_GNUTLS)
./lib/md5.c:#elif defined(USE_OPENSSL)
./lib/vtls/vtls.c:#if defined(USE_OPENSSL) || defined(USE_GNUTLS) || defined(USE_SCHANNEL) || \
./lib/vtls/vtls.c:  defined(USE_DARWINSSL) || defined(USE_POLARSSL) || defined(USE_NSS) || \
./lib/vtls/vtls.c:/* USE_OPENSSL || USE_GNUTLS || USE_SCHANNEL || USE_DARWINSSL || USE_NSS */
./lib/vtls/vtls.c:#elif defined(USE_GNUTLS)
./lib/vtls/vtls.c:#elif defined(USE_NSS)
./lib/vtls/vtls.c:#elif defined(USE_OPENSSL)
./lib/vtls/vtls.c:#if defined(USE_GNUTLS)
./lib/vtls/vtls.c:#if defined(USE_NSS)
./lib/vtls/vtls.c:#if defined(USE_OPENSSL)
./lib/x509asn1.c:#if defined(USE_GSKIT) || defined(USE_NSS) || defined(USE_GNUTLS) || \
./lib/x509asn1.c:#endif /* USE_GSKIT or USE_NSS or USE_GNUTLS or USE_CYASSL or USE_SCHANNEL */
./src/tool_main.c:#ifdef USE_NSS
./src/tool_main.c:#ifdef USE_NSS
./src/tool_metalink.c:#ifdef USE_OPENSSL
./src/tool_metalink.c:#elif defined(USE_GNUTLS_NETTLE)
./src/tool_metalink.c:#elif defined(USE_GNUTLS)
./src/tool_metalink.c:#elif defined(USE_NSS)
./src/tool_metalink.c:#if defined(USE_OPENSSL)
./src/tool_metalink.c:#elif defined(USE_GNUTLS_NETTLE)
./src/tool_metalink.c:#elif defined(USE_GNUTLS)
./src/tool_metalink.c:#elif defined(USE_NSS)
./src/tool_metalink.c:#ifdef USE_NSS
```
All of these _may_ be potential problems when multiple of them is #defined.

vszakats commented:
Metalink is tool-only and it's also dependent on crypto primitives (MD5, SHA1, SHA256). `USE_NSS` together with OpenSSL or GnuTLS will result in a compiler error when Metalink is enabled.
As suspected, when trying to build this combination, this happens:
```
tool_metalink.c:971:6: error: use of undeclared identifier 'nss_context'
  if(nss_context) {
     ^
tool_metalink.c:972:5: warning: implicit declaration of function 'NSS_ShutdownContext' is invalid in C99 [-Wimplicit-function-declaration]
    NSS_ShutdownContext(nss_context);
    ^
  CC       curl-tool_sleep.o
  CC       curl-tool_urlglob.o
tool_metalink.c:972:25: error: use of undeclared identifier 'nss_context'
    NSS_ShutdownContext(nss_context);
                        ^
tool_metalink.c:973:5: error: use of undeclared identifier 'nss_context'
    nss_context = NULL;
    ^
1 warning and 3 errors generated.
```

dscho commented:
> To overcome the 'NSS and SCHANNEL' scenario, an NTLM init function may be a solution, that makes sure that the active backend is correctly initialized.

The cheapest solution would actually be to push `NSS` far, far down in `curl_ntlm_core.c`... then you could keep the rest of your `ntlmmulti` branch as-is.

vszakats commented:
Yes, that's a nice and simple solution. NSS goes to the end as the pickiest backend.

dscho commented:
> USE_NSS together with OpenSSL or GnuTLS will result in a compiler error when Metalink is enabled.

Oy vey. You're right, I had not managed to test this in my setup. Let's keep metalink & MultiSSL for another PR?

dscho commented:
Thank you so much for your hard work, BTW. I really appreciate it.

vszakats commented:
@dscho thank you, and for the exact same to you, too! (also as a GfW user few years back)

vszakats commented:
Let me make a patch for the Metalink issue and decide later. I don't mind either way.

vszakats commented:
Patch for the Metalink issue:
```patch
diff --git a/src/tool_metalink.c b/src/tool_metalink.c
index a20aec14e..a0d10ea12 100644
--- a/src/tool_metalink.c
+++ b/src/tool_metalink.c
@@ -30,6 +30,8 @@
 #  include <fcntl.h>
 #endif
 
+#undef HAVE_NSS_CONTEXT
+
 #ifdef USE_OPENSSL
 #  include <openssl/md5.h>
 #  include <openssl/sha.h>
@@ -50,6 +52,7 @@
 #  define MD5_CTX    void *
 #  define SHA_CTX    void *
 #  define SHA256_CTX void *
+#  define HAVE_NSS_CONTEXT
    static NSSInitContext *nss_context;
 #elif defined(USE_POLARSSL)
 #  include <polarssl/md5.h>
@@ -967,7 +970,7 @@ void clean_metalink(struct OperationConfig *config)
 
 void metalink_cleanup(void)
 {
-#ifdef USE_NSS
+#ifdef HAVE_NSS_CONTEXT
   if(nss_context) {
     NSS_ShutdownContext(nss_context);
     nss_context = NULL;
```

vszakats commented:
Here's some pseudo-PP fu that can help preventing fragile, order dependent `#if` branches in NTLM, Metalink, and other places where it's enough to grab one preferred backend at compile-time. It makes sure this preferred backend gets correctly selected everywhere in the code:
```c
#define USE_OPENSSL
#define USE_NSS
#define USE_GNUTLS
#define USE_WIN32_CRYPT

/* internal header */

#define CURL_VTLS_OPENSSL      1
#define CURL_VTLS_NSS          2
#define CURL_VTLS_GNUTLS       3
#define CURL_VTLS_WIN32_CRYPT  4

#if defined(USE_OPENSSL)
#  define CURL_VTLS_BACKEND  CURL_VTLS_OPENSSL
#elif defined(USE_GNUTLS)
#  define CURL_VTLS_BACKEND  CURL_VTLS_GNUTLS
#elif defined(USE_NSS)
#  define CURL_VTLS_BACKEND  CURL_VTLS_NSS
#elif defined(USE_WIN32_CRYPT)
#  define CURL_VTLS_BACKEND  CURL_VTLS_WIN32_CRYPT
#else
#  undef CURL_VTLS_BACKEND
#endif

/* code */

#include <stdio.h>

int main( void )
{
#if !defined(CURL_VTLS_BACKEND)
  printf("none\n");
#elif (CURL_VTLS_BACKEND == CURL_VTLS_OPENSSL)
  printf("openssl\n");
#elif (CURL_VTLS_BACKEND == CURL_VTLS_GNUTLS)
  printf("gnutls\n");
#elif (CURL_VTLS_BACKEND == CURL_VTLS_NSS)
  printf("nss\n");
#elif (CURL_VTLS_BACKEND == CURL_VTLS_WIN32_CRYPT)
  printf("wincrypt\n");
#endif

  return 0;
}
```

vszakats commented:
Back once again to the 'NSS _and_ SCHANNEL' scenario, the problem with putting NSS _last_ is that we started from the requirement that NSS must come _before_ SCHANNEL, due to issues with the latter. (I forgot about this yesterday.) It'd be more and more useful to know why was this required and is it still relevant today.

jay commented:
Why not initialize all available SSL backends, is there some disadvantage to that? It seems like a can of worms otherwise having to worry about what will be called where (unless I misunderstand)

vszakats commented:
@jay: I'd definitely agree on this, though @dscho's https://github.com/curl/curl/pull/1848#issuecomment-328231542 suggested that initializing NSS interferes with OpenSSL and breaks it. It'd be good to know how/why is this possible.

So we now have two unexpected/mysterious premisses here: Windows Crypto being _inferior_ to OpenSSL/NSS and NSS initialization breaking OpenSSL. Mysterious because all these parts need is basic crypto functions, like MD5/SHA.

kdudka commented:
Sorry for replying late on this!

The reason why we do not initialize NSS on startup is that the initialization is expensive, may cause few seconds delays in some bogus environments, etc. and we do not want to bother users by initialization of NSS unless they can actually take an advantage of it.  Therefore NSS is lazily initialized only when we know that it is actually needed. That is, on the first use of TLS or NTLM.

The problem is that NSS initialization is not thread-safe so a mutex is needed for the lazy initialization to work.  The mutex needs to be created during the global initialization because, when NSS needs to be lazily initialized, it is already too late to create the mutex.

There has been quite a lot of progress on this PR.  I will try it out tomorrow hopefully and provide some feedback...

Thank you for working on this!

kdudka commented:
I have tested 061931de0 (merged with 7e4634cfe as the current upstream HEAD) and I can confirm that it fixes the problem I originally reported.  As I understand it, NSS is now used for NTLM only if NSS is the only enabled crypto library.  It looks like a reasonable fix to me.

vszakats commented:
@kdudka Thank you for your tests, I'm glad it worked. As of the current state of the patch, NSS will be used if none of the higher priority libraries are available. At the moment, OpenSSL and GnuTLS have higher priority.

vszakats commented:
Added the Metalink patch to this PR and rebased it to the latest master. The PR/commit description will need to be updated once it's ready for a merge.

jay commented:
Though we don't know more about the ntlm win32 issue, it appears this is otherwise done?

vszakats commented:
@jay: Yes, correct. I believe it can be addressed in a separate patch once more information is available. Regarding the idea in https://github.com/curl/curl/pull/1848#issuecomment-328244094, it seems unnecessary at this point.

I'm going to edit the commit message to cover the Metalink update and this is done.

vszakats commented:
I've made the subtle edit to the commit message. Let me know if there remains anything to improve.

I was about to add comments to the two other NTLM-related files about the required backend order, but it felt a bit out-of-place. Here it is anyway:

```diff
diff --git a/lib/http_ntlm.c b/lib/http_ntlm.c
index 356a7cda9..59b7a1954 100644
--- a/lib/http_ntlm.c
+++ b/lib/http_ntlm.c
@@ -41,6 +41,17 @@
 #include "vauth/vauth.h"
 #include "url.h"
 
+/* Please keep the SSL backend-specific #if branches in this order:
+
+   1. USE_OPENSSL
+   2. USE_GNUTLS_NETTLE
+   3. USE_GNUTLS
+   4. USE_NSS
+   5. USE_MBEDTLS
+   6. USE_DARWINSSL
+   7. USE_OS400CRYPTO
+   8. USE_WIN32_CRYPTO
+ */
 #if !defined(USE_OPENSSL) && \
     !defined(USE_GNUTLS_NETTLE) && \
     !defined(USE_GNUTLS) && \
diff --git a/lib/vauth/ntlm.c b/lib/vauth/ntlm.c
index c106a68aa..270468bdd 100644
--- a/lib/vauth/ntlm.c
+++ b/lib/vauth/ntlm.c
@@ -44,6 +44,17 @@
 #include "rand.h"
 #include "vtls/vtls.h"
 
+/* Please keep the SSL backend-specific #if branches in this order:
+
+   1. USE_OPENSSL
+   2. USE_GNUTLS_NETTLE
+   3. USE_GNUTLS
+   4. USE_NSS
+   5. USE_MBEDTLS
+   6. USE_DARWINSSL
+   7. USE_OS400CRYPTO
+   8. USE_WIN32_CRYPTO
+ */
 #if !defined(USE_OPENSSL) && \
     !defined(USE_GNUTLS_NETTLE) && \
     !defined(USE_GNUTLS) && \
```

dscho commented:
Thank you so much for your hard work, BTW. I really appreciate it.

> I was about to add comments to the two other NTLM-related files about the required backend order, but it felt a bit out-of-place.

I think this still would make sense. Otherwise, we will forget when changing the code in question (or reviewing changes).

Speaking of reviewing changes: I just went over all the commits with a semi-fresh mind (after not looking at cURL's source code for a few days, kind of simulating how I would stare at those commits in 6 months from now), and the patches look pretty good.

If I may, I would like to ask for two changes, still:

1) The commit message of the first commit does not make it immediately obvious to me why NTLM does not follow the lead of `vtls/*` where the SSL backend is selected at runtime, but instead picks one at compile time. I think I (and probably other readers) would be greatly helped by a commit message like this:

>  ntlm: use strict order for SSL backend #if branches
>
> With the recently introduced MultiSSL support multiple SSL backends can be compiled into cURL That means that now the order of the SSL `#if` branches became significant in the NTLM code.
>
> One option would be to use the same SSL backend as was configured via `curl_global_sslset()`, however, NTLMv2 support would appear to be available only with *some SSL backends*. For example, when eb88d778e (ntlm: Use Windows Crypt API, 2014-12-02) introduced support for NTLMv1 using Windows' Crypt API, it specifically did *not* introduce NTLMv2 support using Crypt API at the same time.
>
> So let's select one specific SSL backend for NTLM support when compiled with multiple SSL backends, using a priority order such that we support NTLMv2 even if only one compiled-in SSL backend can be used for that.

2) When reading the oneliner " metalink: fix NSS issue in MultiSSL builds", it does not immediately become clear to me how this relates to the patch. Maybe some commit message along these lines would help future me (and possibly other readers):

>  metalink: fix NSS issue in MultiSSL builds
>
> In MultiSSL mode (i.e. when more than one SSL backend is compiled in), we cannot use the compile time flag `USE_NSS` as indicator that the NSS backend is in use. As far as Metalink is concerned, the SSL backend is only used for MD5, SHA-1 and SHA-256 calculations, therefore one of the available SSL backends is selected at compile time, in a strict order of preference.
>
> Let's introduce a new `HAVE_NSS_CONTEXT` constant that can be used to determine whether the SSL backend used for Metalink is the NSS backend, and use that to guard the code that wants to de-initialize the NSS-specific data structure.

With these two changes (and [your suggested one](https://github.com/curl/curl/pull/1848#issuecomment-330300644)), I would be very much in favor of this PR, and I would also be *very* excited to see this make it into v7.56.0, still...

The rest of this comment should not be construed as to block this PR.

> /cc @captain-caveman2k

@captain-caveman2k it would be *really* nice to have a proper explanation about NTLMv2 and SSL backends. This is what I found out so far:

While trying to chase down an answer to that question "why does NTLM support differ from SSL backend to SSL backend?" in the commit history and the mailing list archive, I ran across some other conditional clauses in `curl_ntlm_core.h`: https://github.com/curl/curl/blob/c73ebb85374164515eb9df6d619a5822b6568599/lib/curl_ntlm_core.h#L35-L53.

This *may* be the possible explanation for the wincrypt conundrum: it says that `USE_NTLM2SESSION` can only be defined if we have `curl_ssl_md5sum()` (implying that `WIN32_CRYPTO` does not offer that function). That comment was introduced in 97e90d7b8928f49c00ff8109a371e0613821fbc1, and at that time we had no support for Secure Channel (AKA "winssl"), even.

However, looking a bit further, it seems that `USE_WIN32_CRYPTO` is only ever defined by `lib/config-win32.h`, which incidentally is a hand-crafted header to be used when `./configure` is not used to configure the build options, e.g. when using Visual Studio project files from `projects/Windows/`.

This makes it a little less of a concern to me, as `config-win32.h` was not adjusted for MultiSSL support at all, so far. It would even appear that it tries to `#define USE_WIN32_CRYPTO` whenever *neither OpenSSL nor NSS* are configured, meaning that it seems ill-prepared even for the compile axis that uses Secure Channel.

What strikes me as *very* funny, though, is that 8fdf832e5fa39bccab8825f95f7c46b54e12f6fb introduced a default md5sum() function, i.e. if the comment in `curl_ntlm_core.h` is correct, then *any* SSL backend will do to support NTLMv2. And that commit predates the commit that introduced that comment.

So, @captain-caveman2k, what's the deal with NTLMv2 and the SSL backends?

vszakats commented:
@dscho Thanks for writing up these commit message, I'm going to update the current ones with them.

vszakats commented:
Committed the comment [patch](https://github.com/curl/curl/pull/1848#issuecomment-330300644) and updated the commit messages.

dscho commented:
@jay @bagder Could you have a look and merge as you see appropriate? I really would like to see this go in.

@captain-caveman2k I still would love to hear your thoughts on the NTLMv2 matter.

vszakats commented:
Thanks @jay, I got it till the rebase/squash/edit part, the rest I used to to a bit more conservatively, but `--ff-only` and `push upstream` did the trick, and also moved this PR to merged status automatically. Quite nice and a bit unusual, I hope it's okay.

Thanks for everyone the input, patience and assistance!

bagder commented:
Lovely work on this PR everyone, thanks!
